source/geod24/concurrency.d

```diff
@@ -985,6 +986,161 @@ protected:
+                else
+                {
+                    qEnt = new ChannelQueueEntry(&val);
+                    m_txq ~= qEnt;
+                }
+            }
+
+            if (notify_peer)
+            {
+                qEnt.sem.notify();
+            }
+            else if (qEnt)
+            {
+                fiber_was_blocked = true;
+                qEnt.sem.wait();
+            }
+
+            return !(fiber_was_blocked) || qEnt.success;
+        }
+
+        bool read(ref T output)
+        {
+            bool notify_peer, fiber_was_blocked = false;
+            ChannelQueueEntry qEnt;
+
+            synchronized (m_lock)
+            {
+                notify_peer = m_txq.length > 0;
+
+                if (m_buffer.length != 0)
+                {
+                    output = m_buffer[0];
+                    m_buffer = m_buffer.remove(0);
+
+                    if (m_txq.length > 0)
+                    {
+                        qEnt = m_txq[0];
+                        m_txq = m_txq.remove(0);
+
+                        m_buffer ~= *(qEnt.pVal);
+                    }
+                }
+                else if (m_closed)
+                {
+                    return false;
+                }
+                else if (m_txq.length > 0)
+                {
+                    qEnt = m_txq[0];
+                    m_txq = m_txq.remove(0);
+
+                    output = *(qEnt.pVal);
+                }
+                else
+                {
+                    qEnt = new ChannelQueueEntry(&output);
+                    m_rxq ~= qEnt;
+                }
+            }
+
+            if (notify_peer)
+            {
+                qEnt.sem.notify();
+            }
+            else if (qEnt)
+            {
+                fiber_was_blocked = true;
+                qEnt.sem.wait();
+            }
+
+            return !(fiber_was_blocked) || qEnt.success;
+        }
+
+        void close()
+        {
+            synchronized (m_lock)
+            {
+                m_closed = true;
+
+                m_rxq.each!(entry => entry.terminate());
+                m_txq.each!(entry => entry.terminate());
+
+                m_rxq.length = 0;
+                m_txq.length = 0;
+            }
+        }
+
+        ulong length()
+        {
+            return m_buffer.length;
+        }
+
+        private class ChannelQueueEntry
+        {
+            FiberBinarySemaphore sem;
+            T* pVal = null;
+            bool success = true;
+
+            this (T* pVal)
+            {
+                sem = new FiberBinarySemaphore;
+                this.pVal = pVal;
+            }
+
+            void terminate()
+            {
+                success = false;
+                sem.notify();
+            }
+        }
+
+    private:
+        T[] m_buffer;
+        bool m_closed = false;
+        Mutex m_lock;
+
+        ChannelQueueEntry[] m_rxq;
+        ChannelQueueEntry[] m_txq;
+
+    public:
+        immutable ulong size;
```

Comment:
This should have a better name / be documented well. For example, is this an upper bound on the channel size, IOW is it a limit? Does the size mean the number of bytes, or the number of items? And what happens when the size limit is reached?